linkcheck/htmlutil/htmlparse.py

```diff
@@ -0,0 +1,96 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+"""
+HTML parser implemented using Beautiful Soup and html.parser.
+
+USAGE
+
+Two functions are provided, one to make a BeautifulSoup object from markup and
+another to call a handler's callbacks for each element in a BeautifulSoup
+object it can process.
+
+Used callbacks of a handler are:
+
+- Start tag: <tag {attr1:value1, attr2:value2, ..}>
+  def start_element (tag, attrs, text, line, column)
+  @param tag: tag name
+  @type tag: string
+  @param attrs: tag attributes
+  @type attrs: dict
+  @param text: element text
+  @type tag: string
+  @param line: tag line number
+  @type tag: integer
+  @param column: tag column number
+  @type tag: integer
+
+- Start-end tag: <tag {attr1:value1, attr2:value2, ..}/>
+  def start_end_element(tag, attrs, text, line, column):
+  @param tag: tag name
+  @type tag: string
+  @param attrs: tag attributes
+  @type attrs: dict
+  @param text: element text
+  @type tag: string
+  @param line: tag line number
+  @type tag: integer
+  @param column: tag column number
+  @type tag: integer
+
+- End tag: </tag>
+  def end_element (tag)
+  @param tag: tag name
+  @type tag: string
+
+
+EXAMPLE
+
+ # Create a new BeautifulSoup object.
+ soup = htmlutil.htmlparse.make_soup("<html><body>Blubb</body></html>")
+ # Process the soup with the chosen handler as a parameter.
+ htmlutil.htmlparse.proces_soup(handler, soup)
+
+"""
+
+from warnings import filterwarnings
+
+filterwarnings("ignore",
+    message="The soupsieve package is not installed. CSS selectors cannot be used.",
+    category=UserWarning, module="bs4")
+
+from bs4 import BeautifulSoup
+
+
+def make_soup(markup, from_encoding=None):
+    return BeautifulSoup(markup, "html.parser", from_encoding=from_encoding,
+                         multi_valued_attributes=None)
+
+def process_soup(handler, soup):
+    for element in soup.find_all(handler.tag_filter):
+        tag_column = None if element.sourcepos is None \
+            else element.sourcepos + 1
+        if element.is_empty_element:
+            handler.start_end_element(
+                element.name, element.attrs, element.text.strip(),
+                element.sourceline, tag_column
+            )
+        else:
+            handler.start_element(
+                element.name, element.attrs, element.text.strip(),
+                element.sourceline, tag_column
+            )
+            if hasattr(handler, 'end_element'):
+                handler.end_element(element.name)
```

Comment:
Yeah, this is the questionable part all right.

It feels to me that end_element() ought to be called after all of the element's children have been processed.

Do the FormFinder tests not complain about formless inputs where the inputs aren't actually formless?  I suppose warnings emitted during test runs don't actually make the test fail.

Reply:
Yes we don't see the log messages. Will be superseded in the next PR.